scalding-core/src/main/scala/com/twitter/scalding/reducer_estimation/RuntimeReducerEstimator.scala

```diff
@@ -0,0 +1,107 @@
+package com.twitter.scalding.reducer_estimation
+
+import scala.collection.JavaConverters._
+import cascading.flow.FlowStep
+import cascading.tap.{ Tap, CompositeTap }
+import cascading.tap.hadoop.Hfs
+import org.apache.hadoop.mapred.JobConf
+import org.slf4j.LoggerFactory
+
+object RuntimeReducerEstimator {
+  val RuntimePerReducer = "scalding.reducer.estimator.runtime.per.reducer"
+  val defaultRuntimePerReducer = 10 * 60 * 1000 // 10 mins
+
+  /** Get the target bytes/reducer from the JobConf */
+  def getRuntimePerReducer(conf: JobConf): Long = conf.getLong(RuntimePerReducer, defaultRuntimePerReducer)
+
+  def getReduceTimes(history: Seq[FlowStepHistory]): Seq[Seq[Long]] =
+    history.map { h =>
+      h.tasks
+        .filter { t => t.taskType == "REDUCE" && t.status == "SUCCEEDED" && t.finishTime > t.startTime }
+        .map { t => t.finishTime - t.startTime }
+    }
+}
+
+/**
+ * Estimator that uses the input size and a fixed "bytesPerReducer" target.
+ *
+ * Bytes per reducer can be configured with configuration parameter, defaults to 1 GB.
+ */
+trait RuntimeReducerEstimator extends HistoryReducerEstimator {
```

Comment:
Yes, keeping the estimation logic out of the hierarchy makes sense to me. How about something like the following?
```scala
trait RuntimeEstimationScheme {
  def estimateTaskTime(times: Seq[Long]): Option[Long]
  def estimateJobTime(times: Seq[Long]): Option[Long]
}
object MedianBasedEstimationScheme extends RuntimeEstimationScheme {
 ...
}
object MeanBasedEstimationScheme extends RuntimeEstimationScheme {
 ...
}
```
And the estimator classes could be:
 ```scala
class HRavenMedianRuntimeBasedEstimator
  extends BasicRuntimeReducerEstimator(MedianBasedEstimationScheme)
  with HRavenHistoryService 
```
etc.

I'd also suggest using `Basic` instead of `Ordinary` for the names.